### PR TITLE
Native function refactoring

### PR DESCRIPTION
**Pull request by RIg410**
## Motivation

Single interface and expand the functionality of native functions.
Currently there are functions such as write_to_event_store and save_account that require an interpretation context. This creates a single, secure interface for interpreting the native function and interpreter.
The ability to define a native function resolver in the interpreter, which makes it more customizable and reusable.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

Existing tests.

## Related PRs
--

### CODE REVIEW

**Comment by dariorussi** [comment]
@RIg410 thanks for the PR! We had plans to do this work and did not get around to do it yet so it would be great to work together to get this in.
We are however finishing some work right now and we don't have the bandwidth to review and comment and work together to get it in.
We will get to this next week and we will have more talks about native functions!

**Comment by RIg410** [comment]
Hi @dariorussi, is there any chance we can deal with it this week?

**Comment by dariorussi** [comment]
> Hi @dariorussi, is there any chance we can deal with it this week?

we are trying, but please do not expect this to move too fast. We really want to change this part of the code and I will start looking at this today but I am not sure when we are going to get it in.
We will let you know more soon, sorry!

File: language/move-vm/types/src/native_functions/context.rs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    loaded_data::{runtime_types::Type, types::FatType},
+    values::Struct,
+};
+use libra_types::{account_address::AccountAddress, language_storage::ModuleId};
+use move_core_types::{gas_schedule::CostTable, identifier::IdentStr};
+use std::fmt::Write;
+use vm::errors::VMResult;
+
+/// Native function context.
+pub trait NativeContext {
```
Comment (dariorussi) [inline language/move-vm/types/src/native_functions/context.rs:15]: `NativeContext` feels like a very rich API and it's not clear what the reasons are.
You could imagine `NativeContext` to expose only 2 APIs: `CostTable` and `DataStore`.
- `CostTable` is the API every native would need and the **only** API all native functions today need (and those we have in mind)
- `DataStore` is the API that would have to be implemented in order to interact with a data store. `save_data()` and `save_event()` would be the 2 API needed in the Libra case. `DataStore` is tricky because it has to work in concert with the API used for all bytecodes operating on global data.

Moreover `DataStore` seems a bit too special/powerful and we could define a different or parallel way to expose it.
In other words we could imagine not giving `DataStore` to every single native function.

Libra could be made to work with those 2 "concepts" only but we are wondering what your use cases and scenarios are. We understand native functions are a way to expose very tricky behavior and it is not too easy to understand where to draw the line.

Thoughts?

File: language/benchmarks/Cargo.toml [hunk 2]
```diff
@@ -24,6 +24,7 @@ move-core-types = { path = "../move-core/types", version = "0.1.0" }
 move-lang = { path = "../move-lang", version = "0.0.1" }
 move-vm-runtime = { path = "../move-vm/runtime", version = "0.1.0" }
 move-vm-state = { path = "../move-vm/state", version = "0.1.0" }
+move-vm-types = { path = "../move-vm/types", version = "0.1.0" }
```
[the remark this answers is not in the document]
Comment (RIg410) [reply language/benchmarks/Cargo.toml:27]: Nope. I'll fix it.

File: language/move-vm/runtime/src/interpreter.rs [hunk 3]
```diff
@@ -180,7 +177,8 @@ impl<'txn> Interpreter<'txn> {
                     )?;
                     let func = resolver.function_at(fh_idx);
                     if func.is_native() {
-                        self.call_native(&resolver, context, func, vec![])?;
+                        let caller = current_frame.function.module_id();
```
[the remark this answers is not in the document]
Comment (RIg410) [reply language/move-vm/runtime/src/interpreter.rs:174]: I have a case of using a native function where you need to know the calling module.  I have a module similar in functionality to the event module. And the function must know the module from which it is called. Without an explicit indication of the type in code. 
I have a module A, B, C, Store, and a script. The Store module contains a method that will save some resources using calling the native store function. There is a white list of modules from the context of which it is possible to do this, and prohibits the call of Store::store_resources from the script.

File: language/move-vm/types/src/native_functions/account.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    context: &mut impl NativeContext,
+    ty_args: Vec<Type>,
+    mut arguments: VecDeque<Value>,
+) -> VMResult<NativeResult> {
+    let cost = native_gas(context.cost_table(), NativeCostIndex::SAVE_ACCOUNT, 0);
+
+    let address = pop_arg!(arguments, AccountAddress);
+
+    let account_module_id = context.self_module();
+    if address == *account_module_id.address() {
```
[the remark this answers is not in the document]
Comment (RIg410) [reply language/move-vm/types/src/native_functions/account.rs:33]: I agree. This is a redundant function.

File: language/move-vm/types/src/values/values_impl.rs [hunk 5]
```diff
@@ -2152,6 +2156,31 @@ pub mod debug {
 
         Ok(NativeResult::ok(ZERO_GAS_UNITS, vec![]))
     }
+
+    #[allow(unused_variables)]
+    pub fn native_print_stack_trace(
```
[the remark this answers is not in the document]
Comment (RIg410) [reply language/move-vm/types/src/values/values_impl.rs:2161]: I think we should make a single debugging mechanism. 
An interface that will allow you to print arbitrary data of the virtual machine to print the stack, etc., which can be used both for debugging the machine itself, possibly complex native functions, and can also be used in code move as a module Debug. 
Still, I think that a unified interface without special exceptions will be better. Where there's one exception, there'll be a second exception and so on.  We only need to think over the necessary and sufficient interface for native functions, which can be extended if necessary.

File: language/move-vm/types/src/native_functions/context.rs (hunk 1 above)
Comment (RIg410) [reply language/move-vm/types/src/native_functions/context.rs:15]: Yeah. Maybe it's the wrong idea to row all under one API.
I think we should impoverish the input function. (ty_args, arguments) to reduce the number of input parameters. 
And NativeContext should be divided into data sources and context with CostTable and other useful things.
One more important thing is that we would like to be able to deduct gas before executing the body function with the macro gas! 
Two points:
1) the function can be heavy and it might not make sense to spend resources on its execution if there is still not enough gas.
2) The current interface (VMResult<NativeResult>) does not allow to simply use '?' because in this case gas is not deduct.

File: language/move-vm/types/src/native_functions/context.rs [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    /// Gets caller module.
+    fn caller_module(&self) -> Option<&ModuleId>;
+    /// Gets current function module.
+    fn self_module(&self) -> &ModuleId;
+    /// Gets interpreter ref.
```
[the remark this answers is not in the document]
Comment (RIg410) [reply language/move-vm/types/src/native_functions/context.rs:20]: I got it.  

> Just leaving a comment for now with some questions for you.
> We'll be having a meeting about this tomorrow because few people in the team have thought about native functions and had ideas on how to move forward.
> We will be in a better spot then to suggest specific changes.
> 
> Overall a pointer to (or just an explanation on) how you intend to use it may help us understand certain API (as explained in comments).
> In other words we recognize that plugging native functions is pretty vital and it would be great to understand the different use cases.
> 
> Good work btw!

Thank you for your review.

File: language/move-vm/runtime/src/interpreter.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (dariorussi) [reply language/move-vm/runtime/src/interpreter.rs:174]: I have to say I find passing the knowledge of the calling module too much information for a native. But I suppose that depends on what you need to do.
I am curious about your plans to use native functions with the MoveVM. 
I mean which part of the VM are planning to get visibility on and to build what.
I mean, at the end that is knowledge that you can expose explicitly in certain scenarios and if you cannot I am not sure what the developer expectation should be (how do I know what happens and why)

**Comment by dariorussi** [comment]
one more question: what are your plan with natives like the vector API?
is your plan to plug in an entirely different "stdlib" (meaning all the modules that are not just part of the libra ecosystem, but the Move layer as well)?

File: language/move-vm/runtime/src/interpreter.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (sblackshear) [reply language/move-vm/runtime/src/interpreter.rs:174]: I share the concerns about needing to know the calling module. The permissions model of Move is intentionally designed to avoid the need to do a stack walk for security checks.

> I have a module A, B, C, Store, and a script. The Store module contains a method that will save some resources using calling the native store function. There is a white list of modules from the context of which it is possible to do this, and prohibits the call of Store::store_resources from the script.

There are design patterns that can be used to do this without exposing the calling module. The one we usually use is something like
```
native fun store_resources(some_data: u64, capability: &StoreResourcesCapability); 

public fun grant_capability(): StoreResourcesCapability {
  if (// sender is in some whitelist of accepted addresses) {
    move_to_sender(StoreResourcesCapability{})
  }
}


// some other module that you want to allow to call store_resources
module Caller {
  resource T { cap: StoreResourcesCapability }

  public publish() {
    move_to_sender<t> { cap: StoreResources::grant_capability() };
  }

  public call(some_data: u64) {
    StoreResources::store_resources(some_data, &borrow_global<T>.cap)
  }
}

```

This way, it won't be possible for anyone to call `store_resources` unless they can get access to a `&StoreResourcesCapability`. Since `grant_capability` restricts access to the addresses you specify, you can ensure it can't be called  from anywhere you don't expect.

**Comment by olegvg** [comment]
> one more question: what are your plan with natives like the vector API?
> is your plan to plug in an entirely different "stdlib" (meaning all the modules that are not just part of the libra ecosystem, but the Move layer as well)?

Not entirely, but we’re planning to expand/replace a significant part of stdlib with implementation of our own cases. It's aimed to support financial instruments different to libra but with common approaches reflected in Move.

File: language/move-vm/types/src/native_functions/event.rs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        .simple_serialize(&ty)
+        .ok_or_else(|| VMStatus::new(StatusCode::DATA_FORMAT_ERROR))?;
+
+    let count = pop_arg!(arguments, u64);
+    let key = pop_arg!(arguments, Vec<u8>);
+    let guid = EventKey::try_from(key.as_slice())
+        .map_err(|_| VMStatus::new(StatusCode::EVENT_KEY_MISMATCH))?;
+    context
+        .interpreter_mut()
+        .push_event(ContractEvent::new(guid, count, ty.type_tag()?, msg));
```
[the remark this answers is not in the document]
Comment (RIg410) [reply language/move-vm/types/src/native_functions/event.rs]: Done.

File: language/move-vm/types/src/native_functions/context.rs [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    fn interpreter(&self) -> &dyn InterpreterContext;
+    /// Gets interpreter mut ref.
+    fn interpreter_mut(&mut self) -> &mut dyn InterpreterContext;
```
[the remark this answers is not in the document]
Comment (RIg410) [reply language/move-vm/types/src/native_functions/context.rs]: Done.

File: language/move-vm/types/src/native_functions/context.rs [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+use vm::errors::VMResult;
+
+/// Native function context.
+pub trait NativeContext {
+    /// Gets caller module.
+    fn caller_module(&self) -> Option<&ModuleId>;
+    /// Gets current function module.
+    fn self_module(&self) -> &ModuleId;
+    /// Gets interpreter ref.
+    fn interpreter(&self) -> &dyn InterpreterContext;
```
[the remark this answers is not in the document]
Comment (RIg410) [reply language/move-vm/types/src/native_functions/context.rs]: Done.

File: language/move-vm/types/src/native_functions/context.rs [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+use move_core_types::{gas_schedule::CostTable, identifier::IdentStr};
+use std::fmt::Write;
+use vm::errors::VMResult;
+
+/// Native function context.
+pub trait NativeContext {
+    /// Gets caller module.
+    fn caller_module(&self) -> Option<&ModuleId>;
+    /// Gets current function module.
+    fn self_module(&self) -> &ModuleId;
```
[the remark this answers is not in the document]
Comment (RIg410) [reply language/move-vm/types/src/native_functions/context.rs]: Done.

File: language/move-vm/runtime/src/interpreter.rs [hunk 11]
```diff
@@ -245,6 +235,7 @@ impl<'txn> Interpreter<'txn> {
         &mut self,
         resolver: &Resolver,
         context: &mut dyn InterpreterContext,
+        caller_module: Option<&ModuleId>,
```
[the remark this answers is not in the document]
Comment (RIg410) [reply language/move-vm/runtime/src/interpreter.rs]: Done.

File: language/move-vm/runtime/src/interpreter.rs [hunk 12]
```diff
@@ -208,7 +197,8 @@ impl<'txn> Interpreter<'txn> {
                     let func = loader.function_at(func_inst.handle());
                     let ty_args = func_inst.materialize(current_frame.ty_args())?;
                     if func.is_native() {
-                        self.call_native(&resolver, context, func, ty_args)?;
+                        let caller = current_frame.function.module_id();
```
[the remark this answers is not in the document]
Comment (RIg410) [reply language/move-vm/runtime/src/interpreter.rs]: Done.

File: language/move-vm/runtime/src/interpreter.rs [hunk 13]
```diff
@@ -57,23 +50,14 @@ macro_rules! debug_writeln {
...
-
 /// `Interpreter` instances can execute Move functions.
 ///
 /// An `Interpreter` instance is a stand alone execution context for a function.
 /// It mimics execution on a single thread, with an call stack and an operand stack.
 /// The `Interpreter` receives a reference to a data store used by certain opcodes
 /// to do operations on data on chain and a `TransactionMetadata` which is also used to resolve
 /// specific opcodes.
-pub(crate) struct Interpreter<'txn> {
+pub struct Interpreter<'txn> {
```
[the remark this answers is not in the document]
Comment (RIg410) [reply language/move-vm/runtime/src/interpreter.rs]: Done.

File: language/benchmarks/Cargo.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (RIg410) [reply language/benchmarks/Cargo.toml]: Done.

File: language/move-vm/runtime/src/interpreter.rs [hunk 14]
```diff
@@ -125,6 +109,10 @@ impl<'txn> Interpreter<'txn> {
         }
     }
 
+    pub fn gas_schedule(&self) -> &CostTable {
```
[the remark this answers is not in the document]
Comment (RIg410) [reply language/move-vm/runtime/src/interpreter.rs]: Done

File: language/move-vm/runtime/src/interpreter.rs [hunk 15]
```diff
@@ -656,7 +509,11 @@ impl<'txn> Interpreter<'txn> {
     }
 
     #[allow(dead_code)]
-    fn debug_print_stack_trace<B: Write>(&self, buf: &mut B, resolver: &Resolver) -> VMResult<()> {
+    pub fn debug_print_stack_trace<B: Write>(
```
[the remark this answers is not in the document]
Comment (RIg410) [reply language/move-vm/runtime/src/interpreter.rs]: Done.

File: language/move-vm/runtime/src/native_context.rs [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+    loaded_data::{runtime_types::Type, types::FatType},
+    native_functions::context::NativeContext,
+    values::Struct,
+};
+use std::fmt::Write;
+use vm::errors::VMResult;
+
+pub struct FunctionContext<'a, 'txn> {
+    caller_module: Option<&'a ModuleId>,
+    self_module: &'a ModuleId,
```
[the remark this answers is not in the document]
Comment (RIg410) [reply language/move-vm/runtime/src/native_context.rs]: Done.

File: language/move-vm/types/src/native_functions/context.rs [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+};
+use libra_types::{account_address::AccountAddress, language_storage::ModuleId};
+use move_core_types::{gas_schedule::CostTable, identifier::IdentStr};
+use std::fmt::Write;
+use vm::errors::VMResult;
+
+/// Native function context.
+pub trait NativeContext {
+    /// Gets caller module.
+    fn caller_module(&self) -> Option<&ModuleId>;
```
[the remark this answers is not in the document]
Comment (RIg410) [reply language/move-vm/types/src/native_functions/context.rs]: Done

File: language/move-vm/runtime/src/native_context.rs [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+
+    fn self_module(&self) -> &ModuleId {
+        self.self_module
+    }
+
+    fn interpreter(&self) -> &dyn InterpreterContext {
+        self.interpreter_context
+    }
+
+    fn interpreter_mut(&mut self) -> &mut dyn InterpreterContext {
```
[the remark this answers is not in the document]
Comment (RIg410) [reply language/move-vm/runtime/src/native_context.rs]: Done.

File: language/move-vm/runtime/src/native_context.rs [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+impl<'a, 'txn> NativeContext for FunctionContext<'a, 'txn> {
+    fn caller_module(&self) -> Option<&ModuleId> {
+        self.caller_module
+    }
+
+    fn self_module(&self) -> &ModuleId {
+        self.self_module
+    }
+
+    fn interpreter(&self) -> &dyn InterpreterContext {
```
[the remark this answers is not in the document]
Comment (RIg410) [reply language/move-vm/runtime/src/native_context.rs]: Done.

File: language/move-vm/runtime/src/native_context.rs [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+            self_module,
+            interpreter,
+            interpreter_context: context,
+            resolver,
+        }
+    }
+}
+
+impl<'a, 'txn> NativeContext for FunctionContext<'a, 'txn> {
+    fn caller_module(&self) -> Option<&ModuleId> {
```
[the remark this answers is not in the document]
Comment (RIg410) [reply language/move-vm/runtime/src/native_context.rs]: Done.

File: language/move-vm/runtime/src/native_context.rs [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+        self.caller_module
+    }
+
+    fn self_module(&self) -> &ModuleId {
```
[the remark this answers is not in the document]
Comment (RIg410) [reply language/move-vm/runtime/src/native_context.rs]: Done

**Comment by dariorussi** [comment]
if at any point you wanted to share what your Move embedding and usage experience has been, and is, we would not mind to listen.
We are trying to make Move a controlled and expressive language for some onchain operations and it would be good to know the point of frictions with other systems.
Overall we invite you to think about the needs of a calling module and "stack walks" (so to speak).
We believe a model where you provide data and privileges could be trusted more. You do operations over data you receive and you don't go back to look for hints.

But we don't know your model exactly so it's hard to say

Hope you are having fun with your system!

**Comment by sblackshear** [comment]
Let me second @dariorussi's thank you! We are excited to help folks outside of Libra use Move and pleased to get community contributions to the VM, bytecode verifier, etc.
